packages/hint-compat-api/src/utils/parse-supports.ts

```
@@ -0,0 +1,184 @@
+        return !(
+            (token === '(' && isDeclaration(tokens[i + 1])) ||
+            (token === ')' && isDeclaration(tokens[i - 1]))
+        );
+    });
+};
+
+const parseParams = (params: string): DeclarationGroup => {
+    const tokens = tokenizeParams(params);
+
+    const groups: DeclarationGroup[] = [{
+        nodes: [],
+        type: 'and'
+    }];
+
+    for (const token of tokens) {
+        switch (token) {
+            case '(':
+                // Left paren creates a new current group.
+                groups.unshift({
+                    nodes: [],
+                    type: 'and'
+                });
+                // New group is also a node in the previous group.
+                groups[1].nodes.push(groups[0]);
+                break;
+            case ')':
+                // Right paren closes current group.
+                groups.shift();
+                break;
+            case 'and':
+            case 'or':
+            case 'not':
+                // Operator changes current group type.
+                groups[0].type = token;
+                break;
+            default:
+                // Declarations are nodes in the current group.
+                groups[0].nodes.push(token);
+        }
+    }
+
+    if (groups.length !== 1) {
+        throw new Error('Mis-matched parenthesis');
+    }
+
+    return groups[0];
+};
+
+/**
+ * Parse an `@supports` params string into an AST.
+ *
+ * ```css
+ * @supports not ((display: flex) or (display: grid))
+ * ```
+ *
+ * ```js
+ * {
+ *     type: 'not',
+ *     nodes: [
+ *         {
+ *             type: 'or',
+ *             nodes: [
+ *                 { prop: 'display', value: 'flex' },
+ *                 { prop: 'display', value: 'grid' }
+ *             ]
+ *         }
+ *     ]
+ * }
+ * ```
+ */
+export const parseSupports = (params: string): DeclarationGroup | null => {
+    try {
+        return parseParams(params);
+    } catch (e) {
+        return null;
```

Looking at this again I'm thinking I should throw a specific type of error for known invalid states and only filter errors of that type out here. This would avoid silencing unexpected bugs is the parser...